nbconvert/preprocessors/tagremove.py

```diff
@@ -0,0 +1,101 @@
+"""
+Module containing a preprocessor that removes cells if they match
+one or more regular expression.
+"""
+
+# Copyright (c) IPython Development Team.
+# Distributed under the terms of the Modified BSD License.
+
+from traitlets import List, Unicode
+from . import ClearOutputPreprocessor
+
+
+class TagRemovePreprocessor(ClearOutputPreprocessor):
+    """
+    Removes cells from a notebook that have tags that designate they are to be
+    removed prior to exporting the notebook.
+
+    Traitlets:
+    ----------
+    remove_cell_tags: removes cells tagged with these values
+    remove_all_output_tags: removes entire output areas on cells
+                            tagged with these values
+    remove_single_output_tags: removes individual output objects on
+                               outputs tagged with these values
+
+    """
+
+    remove_cell_tags = List(Unicode, default_value=[]).tag(config=True)
+    remove_all_outputs_tags = List(Unicode, default_value=[]).tag(config=True)
+    remove_single_output_tags = List(Unicode, default_value=[]).tag(config=True)
+
+    def check_cell_conditions(self, cell, resources, index):
+        """
+        Checks that a cell has a tag that is to be removed
+
+        Returns: Boolean.
+        True means cell should *not* be removed.
+        """
+
+        # Return true if any of the tags in the cell are removable.
+        return not any([tag in cell.get('metadata', {}).get('tags', [])
+                        for tag in self.remove_cell_tags])
+
+    def preprocess(self, nb, resources):
+        """
+        Preprocessing to apply to each notebook. See base.py for details.
+        """
+        # Skip preprocessing if the list of patterns is empty
+        if not any([self.remove_cell_tags,
+                    self.remove_all_outputs_tags,
+                    self.remove_single_output_tags]):
+            return nb, resources
+
+        # Filter out cells that meet the conditions
+        nb.cells = [self.preprocess_cell(cell, resources, index)[0]
+                    for index, cell in enumerate(nb.cells)
+                    if self.check_cell_conditions(cell, resources, index)]
+
+        return nb, resources
+
+    def preprocess_cell(self, cell, resources, cell_index):
+        """
+        Apply a transformation on each cell. See base.py for details.
+        """
+
+        if (any([tag in cell.get('metadata', {}).get('tags', [])
+                 for tag in self.remove_all_outputs_tags])
+            and cell.cell_type == 'code'):
+            cell.outputs = []
+            cell.execution_count = None
+            # Remove metadata associated with output
+            if 'metadata' in cell:
+                for field in self.remove_metadata_fields:
+                    cell.metadata.pop(field, None)
+        if cell.get('outputs', {}):
```

Comment:
It doesn't really matter to the code, but for the sake of clarity, I'd make the fallback value a list rather than a dict. Or maybe even `None`.